File: backend/model/src/main/java/com/griddynamics/msd365fp/manualreview/model/dfp/CalculatedFields.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.griddynamics.msd365fp.manualreview.model.dfp;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.griddynamics.msd365fp.manualreview.model.DisposabilityCheckServiceResponse;
import com.griddynamics.msd365fp.manualreview.model.jackson.EpochSecondsDateTimeSerializer;
import com.griddynamics.msd365fp.manualreview.model.jackson.FlexibleDateFormatDeserializer;
import com.griddynamics.msd365fp.manualreview.model.jackson.ISOStringDateTimeSerializer;
import lombok.Data;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.List;


@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
public class CalculatedFields implements Serializable {

    // Geo data
    private Boolean matchingOfCountriesForShippingAndIP;
    private Boolean matchingOfCountriesForBillingAndShipping;
    private Boolean matchingOfCountriesForBillingAndIP;
    private List<String> billingCountries;
    private List<String> billingZipCodes;
    private List<String> billingAddresses;
    private BigDecimal distanceToPreviousTransactionIP;

    // Account statistic
    private Long accountAgeInDays;
    private Long activityAgeInDays;
    @JsonDeserialize(using = FlexibleDateFormatDeserializer.class)
    @JsonSerialize(using = ISOStringDateTimeSerializer.class)
    private OffsetDateTime firstTransactionDateTime;
    private Boolean aggregatedEmailConfirmed;
    private String aggregatedEmailDomain;
    private Boolean disposableEmailDomain;
    private List<DisposabilityCheckServiceResponse> disposabilityChecks;

    // Bank data
    private List<String> authBankEventResultCodes;
    private List<String> approveBankEventResultCodes;
    private List<String> declineBankEventResultCodes;

    // Previous transactions statistic
    private Velocity<Long> transactionCount;
    private Velocity<BigDecimal> transactionAmount;
    private Velocity<Long> rejectedTransactionCount;
    private Velocity<BigDecimal> rejectedTransactionAmount;
    private Velocity<Long> failedTransactionCount;
    private Velocity<BigDecimal> failedTransactionAmount;
    private Velocity<Long> successfulTransactionCount;
    private Velocity<BigDecimal> successfulTransactionAmount;
    private Velocity<Long> currentPaymentInstrumentTransactionCount;
    private Velocity<BigDecimal> currentPaymentInstrumentTransactionAmount;
    private Velocity<Long> uniquePaymentInstrumentCount;
    private Velocity<Long> uniqueIPCountries;

    /**
     * Getter for advanced serialization.
     * Json object will contain both representation of DateTime field -
     * timestamp (for SQL queries) and string representation.
     */
    @SuppressWarnings("unused")
    @JsonSerialize(using = EpochSecondsDateTimeSerializer.class)
    public OffsetDateTime getFirstTransactionDateTimeEpochSeconds() {
        return firstTransactionDateTime;
    }
}

File: frontend/src/models/item/purchase/calculated-fields.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { VELOCITIES, VELOCITY_KEYS } from '../../../constants/transaction';
import { CalculatedFieldsDTO } from '../../../data-services/api-services/models/calculated-fields-dto';
import { DisposabilityCheck } from './disposability-check';

export interface TransactionVelocity {
    key: VELOCITY_KEYS;
    name: string;
    isAmount: boolean;
    hour: number;
    day: number;
    week: number;
    lifetime: number | undefined;
    customPlaceholder?: string;
}

export class CalculatedFields {
    matchingOfCountriesForShippingAndIP: boolean = false;

    matchingOfCountriesForBillingAndShipping: boolean = false;

    matchingOfCountriesForBillingAndIP: boolean = false;

    billingCountries: string[] = [];

    billingZipCodes: string[] = [];

    billingAddresses: string[] = [];

    distanceToPreviousTransactionIP?: number;

    accountAgeInDays?: number;

    activityAgeInDays?: number;

    firstTransactionDateTime: string = '';

    firstTransactionDateTimeEpochSeconds: string = '';

    aggregatedEmailConfirmed: boolean = false;

    aggregatedEmailDomain: string = '';

    disposableEmailDomain: boolean = false;

    disposabilityChecks: DisposabilityCheck[] = [];

    authBankEventResultCodes: string[] = [];

    approveBankEventResultCodes: string[] = [];

    declineBankEventResultCodes: string[] = [];

    velocities: TransactionVelocity[] = [];

    fromDTO(calculatedFields: CalculatedFieldsDTO): CalculatedFields {
        const {
            matchingOfCountriesForShippingAndIP,
            matchingOfCountriesForBillingAndShipping,
            matchingOfCountriesForBillingAndIP,
            billingCountries,
            billingZipCodes,
            billingAddresses,
            distanceToPreviousTransactionIP,
            accountAgeInDays,
            activityAgeInDays,
            firstTransactionDateTime,
            firstTransactionDateTimeEpochSeconds,
            aggregatedEmailConfirmed,
            aggregatedEmailDomain,
            disposableEmailDomain,
            disposabilityChecks,
            authBankEventResultCodes,
            approveBankEventResultCodes,
            declineBankEventResultCodes
        } = calculatedFields;

        this.matchingOfCountriesForShippingAndIP = matchingOfCountriesForShippingAndIP;
        this.matchingOfCountriesForBillingAndShipping = matchingOfCountriesForBillingAndShipping;
        this.matchingOfCountriesForBillingAndIP = matchingOfCountriesForBillingAndIP;
        this.billingCountries = billingCountries;
        this.billingZipCodes = billingZipCodes;
        this.billingAddresses = billingAddresses;
        this.distanceToPreviousTransactionIP = distanceToPreviousTransactionIP;
        this.accountAgeInDays = accountAgeInDays;
        this.activityAgeInDays = activityAgeInDays;
        this.firstTransactionDateTime = firstTransactionDateTime;
        this.firstTransactionDateTimeEpochSeconds = firstTransactionDateTimeEpochSeconds;
        this.aggregatedEmailConfirmed = aggregatedEmailConfirmed;
        this.aggregatedEmailDomain = aggregatedEmailDomain;
        this.disposableEmailDomain = disposableEmailDomain;
        this.authBankEventResultCodes = authBankEventResultCodes;
        this.approveBankEventResultCodes = approveBankEventResultCodes;
        this.declineBankEventResultCodes = declineBankEventResultCodes;

        if (Array.isArray(disposabilityChecks)) {
            this.disposabilityChecks = disposabilityChecks.map<DisposabilityCheck>(dc => {
                const dcModel = new DisposabilityCheck();
                return dcModel.fromDTO(dc);
            });
        }

        VELOCITIES.forEach(({
            key, name, isAmount, customPlaceholder
        }) => {
            if (calculatedFields[key]) {
                this.velocities.push({
                    key,
                    name,
                    isAmount,
                    hour: calculatedFields[key].hour,
                    day: calculatedFields[key].day,
                    week: calculatedFields[key].week,
                    lifetime: calculatedFields[key].lifetime,
                    customPlaceholder
                });
            }
        });

        return this;
    }
}
